### PR TITLE
Pest Parser

### PR DESCRIPTION
**Pull request by geralddejong**
Replace the old with the new!

### CODE REVIEW

File: src/build/tenscript/pest_parser.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+fn shape(shape_phase_pair: Pair<Rule>) -> Result<ShapePhase, ParseError> {
+    let mut shape_phase = ShapePhase::default();
+    for pair in shape_phase_pair.into_inner() {
+        match pair.as_rule() {
+            Rule::space_statement => {
+                let mut inner = pair.into_inner();
+                let mark_name = inner.next().unwrap().as_str();
+                let distance_string = inner.next().unwrap().as_str();
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:52]:
```suggestion
                let [mark_name, distance_string] = pair.into_inner().next_chunk().unwrap().map(|p| p.as_str());
```

File: src/build/tenscript/pest_parser.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+fn node(node_pair: Pair<Rule>) -> Result<TenscriptNode, ParseError> {
+    let pair = node_pair.into_inner().next().unwrap();
+    match pair.as_rule() {
+        Rule::face => {
+            let mut inner_pairs = pair.into_inner();
+            let face_name_string = inner_pairs.next().unwrap().as_str();
+            let face_name: FaceName = face_name_string[1..].try_into().unwrap();
+            let node = node(inner_pairs.next().unwrap()).unwrap();
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:111]:
```suggestion
            let [face_name_pair, node_pair] = pair.into_inner().next_chunk().unwrap();
            let face_name_string = face_name_pair.as_str();
            let face_name: FaceName = face_name_string[1..].try_into().unwrap();
            let node = node(node_pair).unwrap();
```

File: src/build/tenscript/pest_parser.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                .map(|post_growth| Box::new(node(post_growth).unwrap()));
+            Ok(TenscriptNode::Grow {
+                forward,
+                scale_factor,
+                post_growth_node,
+            })
+        }
+        Rule::mark => {
+            let mark_name = pair.into_inner().next().unwrap().as_str();
+            Ok(TenscriptNode::Mark { mark_name: mark_name[1..].into() })
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:135]:
```suggestion
            let mark_name = pair.into_inner().next().unwrap().as_str()[1..].into();
            Ok(TenscriptNode::Mark { mark_name })
```

File: src/build/tenscript/pest_parser.rs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            Rule::node => {
+                phase.root = Some(node(pair).unwrap());
+            }
+            _ => unreachable!("build phase: {:?}", pair.as_rule()),
+        }
+    }
+    Ok(phase)
+}
+
+fn node(node_pair: Pair<Rule>) -> Result<TenscriptNode, ParseError> {
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:104]: We can do better than the name node(). build_node()? tenscript_node()?

File: src/build/tenscript/pest_parser.rs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    }
+}
+
+fn scale(scale_pair: Option<Pair<Rule>>) -> f32 {
+    match scale_pair {
+        None => 1.0,
+        Some(scale_pair) => {
+            let scale_string = scale_pair.into_inner().next().unwrap().as_str();
+            scale_string.parse().unwrap()
+        }
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:154]:
```suggestion
                Some(scale_pair) =>
            scale_pair
                .into_inner()
                .next()
                .unwrap()
                .as_str()
                .parse()
                .unwrap(),
```
Matter of preference

File: src/build/tenscript/pest_parser.rs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+use crate::build::tenscript::{BuildPhase, FabricPlan, FaceName, PostShapeOperation, Seed, ShapePhase, ShaperSpec, SurfaceCharacterSpec, TenscriptNode};
+
+#[derive(Parser)]
+#[grammar = "build/tenscript/tenscript.pest"] // relative to src
+struct PestParser;
+
+#[derive(Debug, Clone)]
+pub enum ParseError {
+    Something(String),
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:16]:
```suggestion
    SyntaxError(String),
```

File: src/build/tenscript/pest_parser.rs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            scale_string.parse().unwrap()
+        }
+    }
+}
+
+pub fn parse(source: &str) -> Result<FabricPlan, ParseError> {
+    let mut pairs = PestParser::parse(Rule::fabric_plan, source)
+        .map_err(ParseError::PestError)?;
+    let plan_rule = pairs.next().unwrap();
+    fabric_plan(plan_rule)
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:162]:
```suggestion
    let fabric_plan_pair = pairs.next().unwrap();
    fabric_plan(fabric_plan_pair)
```

Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:162]:
```suggestion
    let fabric_plan_rule = PestParser::parse(Rule::fabric_plan, source)
        .map_err(ParseError::PestError)?
        .next()
        .unwrap();
    fabric_plan(fabric_plan_rule)
```

File: src/build/tenscript/tenscript.pest [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    face_name ~
+    node ~
+    ")"
+}
+
+face_name = {
+    ":A-" |":B-" |
+    ":C-" |":D-" |
+    ":A+" |":B+" |
+    ":C+" |":D+"
```
Comment (mitchelljustin) [inline src/build/tenscript/tenscript.pest:68]: These should omit the ':' prefix, no? In our language, identifiers that start with a capital are automatically atoms.

File: src/build/tenscript/tenscript.pest [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+grow = {
+    "(" ~
+    "grow" ~
+    (integer | forward) ~
```
Comment (mitchelljustin) [inline src/build/tenscript/tenscript.pest:74]: Nice

File: src/build/tenscript/tenscript.pest [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+build = {
+    "(" ~
+    "build" ~
+    seed? ~
+    node? ~
+    ")"
+}
+
+node = {
```
Comment (mitchelljustin) [inline src/build/tenscript/tenscript.pest:52]: I prefer a more specific name for these.

File: src/build/tenscript/tenscript.pest [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+join_statement = {
+    "(" ~
+    "join" ~
+    atom ~
+    ")"
+}
+
+finally_statement = {
```
Comment (mitchelljustin) [inline src/build/tenscript/tenscript.pest:37]:
```suggestion
finally_stmt = {
```
But maybe that's just me

File: src/build/tenscript/pest_parser.rs [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    fn parse_test() {
+        let plans = bootstrap_fabric_plans();
+        for (name, code) in plans.iter() {
+            match parse(code) {
+                Ok(plan) => {
+                    println!("[{name}] Good plan!");
+                    dbg!(plan);
+                }
+                Err(ParseError::PestError(error)) => panic!("[{name}] Error: {error}"),
+                Err(error) => panic!("[{name}] Error: {error:?}"),
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:180]: Alternatively, you could add an `impl Display for ParseError` and match on error type in there. All you'd have to put here is this
```suggestion
                Err(error) => panic!("[{name}] Error: {error}"),
```

Comment (mitchelljustin) [reply src/build/tenscript/pest_parser.rs:180]: To conveniently display errors in other places

File: src/build/tenscript/pest_parser.rs [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+            let node = node(inner_pairs.next().unwrap()).unwrap();
+            Ok(TenscriptNode::Face {
+                face_name,
+                node: Box::new(node),
+            })
+        }
+        Rule::grow => {
+            let mut inner = pair.into_inner();
+            let forward_string = inner.next().unwrap().as_str();
+            let forward = match forward_string.parse() {
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:120]:
```suggestion
            let forward = match inner.next().unwrap().as_str().parse() {
```

File: src/build/tenscript/pest_parser.rs [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+                        ":right-left" => Seed::RightLeft,
+                        ":left" => Seed::Left,
+                        ":right" => Seed::Right,
+                        _ => unreachable!()
+                    };
+            }
+            Rule::node => {
+                phase.root = Some(node(pair).unwrap());
+            }
+            _ => unreachable!("build phase: {:?}", pair.as_rule()),
```
Comment (mitchelljustin) [inline src/build/tenscript/pest_parser.rs:98]:
```suggestion
            _ => unreachable!("build phase: {pair:?}"),
```
For more details

File: src/build/tenscript/mod.rs [hunk 15]
```diff
@@ -1,13 +1,8 @@
...
 
-pub use parser::parse;
-
 use crate::build::tenscript::FaceName::{*};
 
-mod error;
-mod expression;
-mod parser;
-mod scanner;
+mod pest_parser;
```
Comment (mitchelljustin) [inline src/build/tenscript/mod.rs:5]: Just name this "parser"

File: src/build/tenscript/bootstrap.scm [hunk 16]
```diff
@@ -7,7 +7,7 @@
   (build
     (seed :right-left)
     (branch
-      (face :A+ (grow "XXX"))
+      (face :A+ (grow 3))
```
Comment (mitchelljustin) [inline src/build/tenscript/bootstrap.scm:10]: Intentional?

File: src/build/tenscript/tenscript.pest (hunk 8 above)
Comment (geralddejong) [reply src/build/tenscript/tenscript.pest:68]: Ok

File: src/build/tenscript/tenscript.pest (hunk 11 above)
Comment (geralddejong) [reply src/build/tenscript/tenscript.pest:37]: It is just you

File: src/build/tenscript/bootstrap.scm (hunk 16 above)
Comment (geralddejong) [reply src/build/tenscript/bootstrap.scm:10]: Yes

File: src/build/tenscript/parser.rs [hunk 17]
```diff
@@ -1,244 +1,185 @@
...
-                    _ => return Err(Mismatch { expected: "face | grow | mark | branch | scale", expression: expression.clone() }),
-                }
-            }
-            Ok(TenscriptNode::Grow { scale_factor, forward, post_growth_node })
+fn node(node_pair: Pair<Rule>) -> Result<BuildNode, ParseError> {
+    let pair = node_pair.into_inner().next().unwrap();
+    match pair.as_rule() {
+        Rule::face => {
+            let [face_name_pair, node_pair] = pair.into_inner().next_chunk().unwrap();
+            let face_name: FaceName = face_name_pair.as_str().try_into().unwrap();
```
Comment (mitchelljustin) [inline src/build/tenscript/parser.rs:109]:
```suggestion
            let face_name = face_name_pair.as_str().try_into().unwrap();
```

**Review by mitchelljustin — APPROVED** [review]

**Pull request merged**